File: packages/nodejs/ios/Plugin/Nodejs.swift
import Foundation
import Capacitor
import NodejsPluginNative

@objc public class Nodejs: NSObject {
    private static let channelEvents = "_EVENTS_"
    private static let channelSystem = "_SYSTEM_"
    private static let systemMessageReady = "ready-for-app-events"

    private let config: NodejsConfig
    private weak var plugin: NodejsPlugin?

    private var ready = false
    private var started = false

    init(plugin: NodejsPlugin, config: NodejsConfig) {
        self.config = config
        self.plugin = plugin
        super.init()
        NodeRunner.setMessageHandler { [weak self] channelName, message in
            self?.handleMessage(channelName: channelName, message: message)
        }
        if let dataDirPath = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first {
            NodeRunner.registerDataDirPath(dataDirPath)
        }
    }

    @objc public func isReady() -> Bool {
        return ready
    }

    @objc public func send(_ options: SendOptions) throws {
        guard ready else {
            throw CustomError.nodeNotReady
        }
        guard JSONSerialization.isValidJSONObject(options.args) else {
            throw NSError(domain: "NodejsPlugin", code: 0, userInfo: [NSLocalizedDescriptionKey: "args must be JSON-serializable."])
        }
        let payload = try JSONSerialization.data(withJSONObject: options.args)
        let envelope: [String: Any] = [
            "event": options.eventName,
            "payload": String(data: payload, encoding: .utf8) ?? "[]"
        ]
        let envelopeData = try JSONSerialization.data(withJSONObject: envelope)
        guard let message = String(data: envelopeData, encoding: .utf8) else {
            return
        }
        NodeRunner.sendMessage(Nodejs.channelEvents, message: message)
    }

    @objc public func start(_ options: StartOptions, completion: @escaping (_ error: Error?) -> Void) {
        objc_sync_enter(self)
        defer {
            objc_sync_exit(self)
        }
        guard !started else {
            completion(CustomError.nodeAlreadyRunning)
            return
        }
        started = true
        let thread = Thread { [self] in
            do {
                let projectPath = try self.getProjectPath()
                let builtinModulesPath = self.getBuiltinModulesPath()
                let scriptPath = try self.resolveScriptPath(projectPath: projectPath, script: options.script)
                for (key, value) in options.env {
                    setenv(key, value, 1)
                }
                setenv("TMPDIR", NSTemporaryDirectory(), 1)
                var arguments = ["node", scriptPath]
                arguments.append(contentsOf: options.args)
                var nodePath = projectPath
                if let builtinModulesPath = builtinModulesPath {
                    nodePath += ":" + builtinModulesPath
                }
                completion(nil)
                NodeRunner.startEngine(withArguments: arguments, nodePath: nodePath)
            } catch {
                self.started = false
                completion(error)
            }
        }
        thread.name = "NodejsPlugin"
        thread.stackSize = 2 * 1024 * 1024
        thread.start()
    }

    private func handleMessage(channelName: String, message: String) {
        if channelName == Nodejs.channelSystem {
            if message == Nodejs.systemMessageReady {
                ready = true
                plugin?.notifyReadyListeners()
            }
        } else if channelName == Nodejs.channelEvents {
            guard let envelopeData = message.data(using: .utf8),
                  let envelope = try? JSONSerialization.jsonObject(with: envelopeData) as? [String: Any],
                  let eventName = envelope["event"] as? String else {
                return
            }
            var args: [Any] = []
            if let payload = envelope["payload"] as? String, let payloadData = payload.data(using: .utf8) {
                args = (try? JSONSerialization.jsonObject(with: payloadData) as? [Any]) ?? []
            }
            plugin?.notifyMessageListeners(MessageEvent(eventName: eventName, args: args))
        }
    }

    private func getProjectPath() throws -> String {
        let projectPath = Bundle.main.bundlePath + "/public/" + config.nodeDir
        guard FileManager.default.fileExists(atPath: projectPath) else {
            throw CustomError.projectNotFound
        }
        return projectPath
    }

    private func getBuiltinModulesPath() -> String? {
        #if SWIFT_PACKAGE
        return Bundle.module.path(forResource: "builtin_modules", ofType: nil)
        #else
        return Bundle(for: Nodejs.self).path(forResource: "builtin_modules", ofType: nil)
            ?? Bundle.main.path(forResource: "builtin_modules", ofType: nil)
        #endif
    }

    private func resolveScriptPath(projectPath: String, script: String?) throws -> String {
        var scriptFile = script
        if scriptFile == nil {
            scriptFile = getMainFromPackageJson(projectPath: projectPath)
        }
        if scriptFile == nil {
            scriptFile = "index.js"
        }
        let scriptPath = projectPath + "/" + (scriptFile ?? "index.js")
        guard FileManager.default.fileExists(atPath: scriptPath) else {
            throw CustomError.scriptNotFound
        }
        return scriptPath
    }

    private func getMainFromPackageJson(projectPath: String) -> String? {
        let packageJsonPath = projectPath + "/package.json"
        guard let packageJsonData = FileManager.default.contents(atPath: packageJsonPath),
              let packageJson = try? JSONSerialization.jsonObject(with: packageJsonData) as? [String: Any] else {
            return nil
        }
        return packageJson["main"] as? String
    }
}

File: packages/nodejs/android/src/main/java/io/capawesome/capacitorjs/plugins/nodejs/Nodejs.java
package io.capawesome.capacitorjs.plugins.nodejs;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageInfo;
import android.content.res.AssetManager;
import android.system.Os;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.JSArray;
import com.getcapacitor.Logger;
import io.capawesome.capacitorjs.plugins.nodejs.classes.CustomExceptions;
import io.capawesome.capacitorjs.plugins.nodejs.classes.events.MessageEvent;
import io.capawesome.capacitorjs.plugins.nodejs.classes.options.SendOptions;
import io.capawesome.capacitorjs.plugins.nodejs.classes.options.StartOptions;
import io.capawesome.capacitorjs.plugins.nodejs.interfaces.EmptyCallback;
import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.json.JSONObject;

public class Nodejs {

    private static final String ASSETS_BUILTIN_MODULES_DIR = "builtin_modules";
    private static final String BASE_DIR = "capawesome_nodejs";
    private static final String CHANNEL_EVENTS = "_EVENTS_";
    private static final String CHANNEL_SYSTEM = "_SYSTEM_";
    private static final String SHARED_PREFS = "CapawesomeNodejsPlugin";
    private static final String SHARED_PREFS_APK_LAST_UPDATE_TIME = "apkLastUpdateTime";
    private static final String SYSTEM_MESSAGE_PAUSE = "pause";
    private static final String SYSTEM_MESSAGE_READY = "ready-for-app-events";
    private static final String SYSTEM_MESSAGE_RESUME = "resume";

    @Nullable
    private static Nodejs instance;

    static {
        System.loadLibrary("node");
        System.loadLibrary("capacitor-nodejs");
    }

    @NonNull
    private final NodejsConfig config;

    @NonNull
    private final Context context;

    @NonNull
    private final NodejsPlugin plugin;

    private volatile boolean ready = false;

    private volatile boolean started = false;

    public Nodejs(@NonNull NodejsPlugin plugin, @NonNull Context context, @NonNull NodejsConfig config) {
        this.config = config;
        this.context = context;
        this.plugin = plugin;
        Nodejs.instance = this;
    }

    public boolean isReady() {
        return ready;
    }

    public void send(@NonNull SendOptions options) throws Exception {
        if (!ready) {
            throw CustomExceptions.NODE_NOT_READY;
        }
        JSONObject envelope = new JSONObject();
        envelope.put("event", options.getEventName());
        envelope.put("payload", options.getArgs().toString());
        sendMessageToNodeChannel(CHANNEL_EVENTS, envelope.toString());
    }

    public synchronized void start(@NonNull StartOptions options, @NonNull EmptyCallback callback) {
        if (started) {
            callback.error(CustomExceptions.NODE_ALREADY_RUNNING);
            return;
        }
        started = true;
        Thread thread = new Thread(() -> {
            try {
                File projectDir = new File(getBaseDir(), "nodejs-project");
                File builtinModulesDir = new File(getBaseDir(), ASSETS_BUILTIN_MODULES_DIR);
                copyNodeProjectFromApk(projectDir, builtinModulesDir);
                File scriptFile = resolveScriptFile(projectDir, options.getScript());
                for (Map.Entry<String, String> entry : options.getEnv().entrySet()) {
                    Os.setenv(entry.getKey(), entry.getValue(), true);
                }
                Os.setenv("TMPDIR", context.getCacheDir().getAbsolutePath(), true);
                registerNodeDataDirPath(context.getFilesDir().getAbsolutePath());
                List<String> args = new ArrayList<>();
                args.add("node");
                args.add(scriptFile.getAbsolutePath());
                args.addAll(options.getArgs());
                String nodePath = projectDir.getAbsolutePath() + ":" + builtinModulesDir.getAbsolutePath();
                callback.success();
                startNodeWithArguments(args.toArray(new String[0]), nodePath, true);
            } catch (Exception exception) {
                started = false;
                callback.error(exception);
            }
        });
        thread.start();
    }

    public void handleOnPause() {
        if (ready) {
            sendMessageToNodeChannel(CHANNEL_SYSTEM, SYSTEM_MESSAGE_PAUSE);
        }
    }

    public void handleOnResume() {
        if (ready) {
            sendMessageToNodeChannel(CHANNEL_SYSTEM, SYSTEM_MESSAGE_RESUME);
        }
    }

    @SuppressWarnings("unused")
    private static void onMessageReceived(@NonNull String channelName, @NonNull String message) {
        Nodejs instance = Nodejs.instance;
        if (instance == null) {
            return;
        }
        try {
            instance.handleMessage(channelName, message);
        } catch (Exception exception) {
            Logger.error(NodejsPlugin.TAG, "Failed to handle message from the Node.js runtime.", exception);
        }
    }

    private void handleMessage(@NonNull String channelName, @NonNull String message) throws Exception {
        if (CHANNEL_SYSTEM.equals(channelName)) {
            if (SYSTEM_MESSAGE_READY.equals(message)) {
                ready = true;
                plugin.notifyReadyListeners();
            }
        } else if (CHANNEL_EVENTS.equals(channelName)) {
            JSONObject envelope = new JSONObject(message);
            String eventName = envelope.getString("event");
            JSArray args = new JSArray(envelope.optString("payload", "[]"));
            plugin.notifyMessageListeners(new MessageEvent(eventName, args));
        }
    }

    private void copyNodeProjectFromApk(@NonNull File projectDir, @NonNull File builtinModulesDir) throws Exception {
        AssetManager assetManager = context.getAssets();
        String projectAssetsPath = "public/" + config.getNodeDir();
        String[] projectAssets = assetManager.list(projectAssetsPath);
        if (projectAssets == null || projectAssets.length == 0) {
            throw CustomExceptions.PROJECT_NOT_FOUND;
        }
        SharedPreferences sharedPreferences = context.getSharedPreferences(SHARED_PREFS, Context.MODE_PRIVATE);
        long previousLastUpdateTime = sharedPreferences.getLong(SHARED_PREFS_APK_LAST_UPDATE_TIME, 0);
        PackageInfo packageInfo = context.getPackageManager().getPackageInfo(context.getPackageName(), 0);
        long lastUpdateTime = packageInfo.lastUpdateTime;
        if (lastUpdateTime != previousLastUpdateTime || !projectDir.exists()) {
            deleteRecursively(projectDir);
            deleteRecursively(builtinModulesDir);
            copyAssetDirectory(assetManager, projectAssetsPath, projectDir);
            copyAssetDirectory(assetManager, ASSETS_BUILTIN_MODULES_DIR, builtinModulesDir);
            sharedPreferences.edit().putLong(SHARED_PREFS_APK_LAST_UPDATE_TIME, lastUpdateTime).apply();
        }
    }

    @NonNull
    private File resolveScriptFile(@NonNull File projectDir, @Nullable String script) throws Exception {
        String scriptPath = script;
        if (scriptPath == null) {
            scriptPath = getMainFromPackageJson(projectDir);
        }
        if (scriptPath == null) {
            scriptPath = "index.js";
        }
        File scriptFile = new File(projectDir, scriptPath);
        if (!scriptFile.exists()) {
            throw CustomExceptions.SCRIPT_NOT_FOUND;
        }
        return scriptFile;
    }

    @Nullable
    private String getMainFromPackageJson(@NonNull File projectDir) {
        try {
            File packageJsonFile = new File(projectDir, "package.json");
            if (!packageJsonFile.exists()) {
                return null;
            }
            byte[] bytes = new byte[(int) packageJsonFile.length()];
            try (InputStream inputStream = new java.io.FileInputStream(packageJsonFile)) {
                int read = inputStream.read(bytes);
                if (read != bytes.length) {
                    return null;
                }
            }
            JSONObject packageJson = new JSONObject(new String(bytes, StandardCharsets.UTF_8));
            return packageJson.optString("main", null);
        } catch (Exception exception) {
            return null;
        }
    }

    @NonNull
    private File getBaseDir() {
        return new File(context.getFilesDir(), BASE_DIR);
    }

    private void copyAssetDirectory(@NonNull AssetManager assetManager, @NonNull String assetPath, @NonNull File targetDir)
        throws Exception {
        String[] children = assetManager.list(assetPath);
        if (children == null || children.length == 0) {
            copyAssetFile(assetManager, assetPath, targetDir);
        } else {
            if (!targetDir.exists() && !targetDir.mkdirs()) {
                throw new Exception("Failed to create directory: " + targetDir.getAbsolutePath());
            }
            for (String child : children) {
                copyAssetDirectory(assetManager, assetPath + "/" + child, new File(targetDir, child));
            }
        }
    }

    private void copyAssetFile(@NonNull AssetManager assetManager, @NonNull String assetPath, @NonNull File targetFile) throws Exception {
        try (InputStream inputStream = assetManager.open(assetPath); OutputStream outputStream = new FileOutputStream(targetFile)) {
            byte[] buffer = new byte[8192];
            int length;
            while ((length = inputStream.read(buffer)) > 0) {
                outputStream.write(buffer, 0, length);
            }
        }
    }

    private void deleteRecursively(@NonNull File file) {
        if (file.isDirectory()) {
            File[] children = file.listFiles();
            if (children != null) {
                for (File child : children) {
                    deleteRecursively(child);
                }
            }
        }
        file.delete();
    }

    private static native void registerNodeDataDirPath(String dataDir);

    private static native void sendMessageToNodeChannel(String channelName, String msg);

    private static native int startNodeWithArguments(String[] arguments, String nodePath, boolean redirectOutputToLogcat);
}
